File: ClientLibs/Haskell/Commutator.hs
module Commutator where
import Network.Socket hiding (send, sendTo, recv, recvFrom)
import Network.Socket.ByteString (send, recv)
import Tools
import qualified Data.ByteString.Char8 as B8


----- State =  Connected or not ? -----
type State = Int
type ServerInfo = AddrInfo
data Commutator  = Commutator Socket ServerInfo State 

endLine :: String
endLine = "\r\n"

getCommutator :: String -> String -> IO Commutator
getCommutator host port = withSocketsDo $ do 
			addrInfo <- getAddrInfo Nothing (Just host) (Just port)
			let server = head addrInfo
			sock <- socket (addrFamily server) Stream defaultProtocol
			return $ Commutator sock server 0

connectCommutator :: Commutator -> IO Commutator
connectCommutator comm = do
			let Commutator socket server _ = comm
			connect socket (addrAddress server)
			return $ Commutator socket server 1

sendAll :: Socket -> String -> IO Int
sendAll socket message = do
			send socket $ B8.pack (message ++ endLine)

recvAll :: Socket -> IO String
recvAll socket = do
			out <- recv socket 1024
			return $  erase (B8.unpack out) endLine

messageExchange :: String -> Commutator -> IO String
messageExchange message comm =  if (state == 1) then do
									sendAll socket message
									recvAll socket
							    else return ""
							    where Commutator socket _ state  =  comm


disconnectCommutator :: String -> Commutator -> IO Commutator
disconnectCommutator lastmessage comm = if (state == 1) then do
											sendAll socket lastmessage
											sClose socket
											return $ Commutator socket server 0
									    else return comm
							    		where Commutator socket server state  =  comm

handShake :: String ->  Commutator -> IO String
handShake hshkmsg comm = if(state == 1) then do
							sendAll socket hshkmsg
							recvAll socket
						 else return ""
						where Commutator socket _ state  =  comm

--main = do
--	comm <- connectCommutator $ getCommutator "192.168.0.2" "2001"
--	msg <- handShake "HI" comm
--	print msg

File: ClientLibs/Python/commutator.py
import socket

class Commutator:
    end_line = "\r\n"

    def __init__(self, hostname = "localhost", port = "1234"):
        self.port = int(port)
        self.host = hostname
        self.connected = False
        self.buffer_size = 1024
        self.buffer = []
        self.socket = socket.socket(socket.AF_INET,socket.SOCK_STREAM)

        
    def up_connection(self):
        try:        
            self.socket.connect((self.host,self.port))
            self.connected = True
            return True
        except socket.error:
            print "Error, problems with connection to host!"
            return False

    def send_all(self,msg):
        msg += self.end_line
        try:
            if (self.connected):
                self.socket.sendall(msg) 
        except:
            print "Error, problems with sending!"

    def recv_all(self):
        try:
            out_message = ""
            condition = True
            while condition :
                self.buffer = self.socket.recv(self.buffer_size)
                if(len(self.buffer) > 0):
                    out_message += self.buffer
                    condition = not self.check_message_intefrity(str(out_message))
                else:
                    raise socket.error

            index = out_message.find(self.end_line)
            if (index != -1):
                out_message = out_message.replace(self.end_line,'')
            return str(out_message)
                
        except socket.error:
            print "Error, problems with receiving!"
            return ""

    def check_message_intefrity(self,msg):
        j = len(msg) - 1
        for i in range(len(self.end_line) - 1, -1 , -1):
            if(j < 0):
                return False
            if(msg[j] != self.end_line[i]): 
                return False
            j -= 1
        return True


    def exchange(self,request):
        if(self.connected):
            self.send_all(request)
            return self.recv_all()
        else:
            return ""

    def hand_shake(self,hand_shake_msg):
        if(self.connected):
            self.send_all(hand_shake_msg)
            #print self.recv_all()
            return True
        else:
            return False

    def down_connection(self,last_msg):
        if(self.connected):
            self.send_all(last_msg)
            self.connected = False
            self.socket.close()
            return True
        else:
            return False


#cm = Commutator("192.168.0.2","2001")
#cm.up_connection()
#cm.hand_shake("0")
#cm.down_connection("EOG")

